pkg/generator/service_account_loader.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package generator
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	certutil "k8s.io/client-go/util/cert"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrl "sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+const (
+	tokenKey    = "token"
+	caCert      = "ca.crt"
+	saTokenType = "kubernetes.io/service-account-token"
+)
+
+// ServiceAccountLoader allows the construction of a k8s client from a Service Account
+type ServiceAccountLoader struct {
+	client client.Client // Used to load service accounts and their secrets.
+}
+
+// NewServiceAccountLoader creates a new ServiceAccountLoader
+func NewServiceAccountLoader(client client.Client) *ServiceAccountLoader {
+	return &ServiceAccountLoader{client}
+}
+
+// Client returns a new k8s client for a Service Account
+func (s *ServiceAccountLoader) Client(ctx context.Context, saName, saNamespace string) (client.Client, error) {
+	config, err := s.restConfig(ctx, saName, saNamespace)
+	if err != nil {
+		return nil, err
+	}
+
+	return client.New(config, client.Options{})
+}
+
+func (s *ServiceAccountLoader) restConfig(ctx context.Context, saName, saNamespace string) (*rest.Config, error) {
+	cfg, err := ctrl.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	token, cert, err := s.serviceAccountCredentials(ctx, saName, saNamespace)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := certutil.NewPoolFromBytes(cert); err != nil {
+		return nil, fmt.Errorf("expected to load root CA config, but got err: %v", err)
+	}
+
+	return &rest.Config{
+		Host: cfg.Host,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData: cert,
+		},
+		BearerToken: string(token),
+	}, nil
+}
+
+func (s *ServiceAccountLoader) serviceAccountCredentials(ctx context.Context, name, namespace string) ([]byte, []byte, error) {
```

Comment:
we ll want to change this to use TokenRequest API (k8s 1.24 break this implementation):

- https://github.com/vmware-tanzu/carvel-kapp-controller/issues/687
- beginnings of kc change: https://github.com/vmware-tanzu/carvel-kapp-controller/pull/695

Reply:
cc @joe-kimmel-vmw @neil-hickey

Reply:
after https://github.com/vmware-tanzu/carvel-secretgen-controller/pull/67 is merged you can pull-rebase to get some test failures when you access SA creds the 'old' way.

Reply:
Looks like there's no issue for secretgen capturing what you did in #67 @joe-kimmel-vmw.  For now I think we'll copy over the satoken package and utilise it in our service_account_loader as part of this PR!

Reply:
So the latest commits now include a version of the `satoken` package.  Initially it was failing in the e2e tests due to the serviceaccount being deleted and then recreated but the cache could not distinguish here. There are two things I can see us doing:
1. Use a cache key that is universal across all serviceaccounts.  We could get the SA and then use its UID perhaps
2. Use the TokenReview API which takes it into account.

Currently the implementation goes for 2) but let us know if yous have a preference.

Reply:
Also note that we believe Kapp controller will have the same issue so the fix could be ported over.

Reply:
I have created an issue in kapp with steps that I believe will trigger the same issue https://github.com/vmware-tanzu/carvel-kapp-controller/issues/701#issue-1248798176